java/src/jmri/util/TypeConversionUtil.java

```diff
@@ -0,0 +1,321 @@
+     *   <li>BigDecimal</li>
+     *   <li>Float</li>
+     *   <li>Double</li>
+     * </ul>
+     * @param object the object to check
+     * @return true if the object is an object that is either an integer or a
+     * float, false otherwise
+     */
+    public static boolean isFloatingNumber(Object object) {
+        return isIntegerNumber(object)
+                || (object instanceof java.math.BigDecimal)
+                || (object instanceof Float)
+                || (object instanceof Double);
+    }
+    
+    /**
+     * Is this object a String?
+     * @param object the object to check
+     * @return true if the object is a String, false otherwise
+     */
+    public static boolean isString(Object object) {
+        return object instanceof String;
+    }
+    
+    
+    private static boolean convertStringToBoolean(@Nonnull String str, boolean do_i18n) {
+        // try to parse the string as a number
+        try {
+            double number;
+            if (do_i18n) {
+                number = IntlUtilities.doubleValue(str);
+            } else {
+                number = Double.parseDouble(str);
+            }
+//                System.err.format("The string: '%s', result: %1.4f%n", str, (float)number);
+            return ! ((-0.5 < number) && (number < 0.5));
+        } catch (NumberFormatException | ParseException ex) {
+            log.debug("The string '{}' cannot be parsed as a number", str);
+        }
+
+//            System.err.format("The string: %s, %s%n", str, value.getClass().getName());
+        String patternString = "^0(\\.0+)?$";
+        Pattern pattern = Pattern.compile(patternString, Pattern.CASE_INSENSITIVE);
+        Matcher matcher = pattern.matcher(str);
+        if (matcher.matches()) {
+//                System.err.format("The string: '%s', result: %b%n", str, false);
+            return false;
+        }
+//            System.err.format("The string: '%s', result: %b%n", str, !str.isEmpty());
+        return !str.isEmpty();
+    }
+    
+    /**
+     * Convert a value to a boolean.
+     * <P>
+     * Rules:
+     * null is converted to false
+     * empty string is converted to false
+     * "0" string is converted to false
+     * "0.000" string is converted to false, if the number of decimals is &gt; 0
+     * empty map is converted to false
+     * empty collection is converted to false
+     * An integer number is converted to false if the number is 0
+     * A floating number is converted to false if the number is -0.5 &lt; x &lt; 0.5
+     * Everything else is converted to true
+     * <P>
+     * For objects that implement the Reportable interface, the value is fetched
+     * from the method toReportString().
+     * 
+     * @param value the value to convert
+     * @param do_i18n true if internationalization should be done, false otherwise
+     * @return the boolean value
+     */
+    public static boolean convertToBoolean(@CheckForNull Object value, boolean do_i18n) {
```

Comment:
Could you please add some discussion as to when this type-not-safe method should be used, and when not?  Java has spent a decade moving toward type-safety via generics, multiple methods, etc.  So this is a bit off to the side.

For example, another approach would be have parallel methods with arguments typed Map, Collection, etc instead of decoding the type. 

Is there some motivation for actually carrying variables of Object type around?

Reply:
Expressions in LogixNG evaluates to a specific type, for example the ExpressionTurnout checks if a turnout has a specific state and returns true or false. And the StringExpressionMemory reads a memory and returns its value as a string. If a LocoNet device supports sending analog values, for example an axle counter, an analog expression could read that analog value and return it as such.

But what if you want to do some calculations? You might have stored the previous value from the axle counter in a memory and you now want to add the memory value to the current value of the axle counter. For this purpose, there is the expression Formula, which takes a number of expressions and a formula and calculates the formula and returns the result.

But what if you want to calculate some things that has different types? If you have axle counter A from Digitrax that sends the value as a string and you have axle counter B from Roco that sends the value as an integer? How do you add an integer to a string?

The expression Formula can handle this. Its child expressions can be of any type, for example analog expression or string expression, and you can tell Formula to convert the expression E2 to an integer and then add it to expression E1.

This example shows a ConditionalNG that has an action that reads an analog value and then sets an analog value. It reads the result of a formula, which in this case is `E1 + int(E2)`. E1 is the analog value of memory IM_91 and E2 is the string value of memory IM_92. `int(x)` do the conversion from string to integer. It then puts the sum in the memory IM_100.
![LogixNG_Formula](https://user-images.githubusercontent.com/20255317/99602154-466fd480-2a01-11eb-9c1f-ce46ab752434.png)

This is not limited to simple addition or multiplication. The expression Formula supports different functions, for example `sin(x)`, which may be useful if you want to take the hour of the fast clock, convert it to a sine wave and use it to set the intensity of the light in the room to get the impression of day and night in the room.

Due to this, I don't know in advance which type of data Formula will handle. The expression E1 may be a string, expression E2 an analog value and expression E3 may be a boolean. You might want a formula like this:

`(E1 == "Big yard") && ((E2 < 32) || E3)`

This is a valid formula if E1 is a string, E2 an double and E3 a boolean.

Reply:
Thanks for the detail.  What does the expression
```
(E1 == "Big yard") && ((E2 < 32) || E3)
```
do if  E1 is a double, E2 a string and E3 a string?

I'd expect that the E1 comparison would always fail, so this would evaluated to false always.  Setting that aside, there's probably rules for converting a string to a Boolean, so the E3 term can be handled.  But what happens to `E2 < 32`?  Can you guarantee that exactly one of `E2<32`, `E2==32`, and `E2>32` is true?

Reply:
I try to follow the rules of Jython since I expect that that is what most users are familiar with. It was maybe a year ago I wrote the formula expression so I don't remember the details, but I tried different expressions in Jython and did the same thing in Formula.

`print 10 + "Hello"` causes an error in Jython, so I do the same in Formula.

`print 10 < "0"` returns true and `print 10 > "0"` returns false, and Formula does the same.

Reply:
> Can you guarantee that exactly one of `E2<32`, `E2==32`, and `E2>32` is true?

The comparision method is here:
https://github.com/danielb987/JMRI/blob/26e005cc5808aed457401fd79889de673ec1de47/java/src/jmri/jmrit/logixng/util/parser/ExpressionNodeComparingOperator.java#L39
The method `calculate()` at lines 39 - 107 does the evaluation.

The arithmetic operators are here:
https://github.com/danielb987/JMRI/blob/26e005cc5808aed457401fd79889de673ec1de47/java/src/jmri/jmrit/logixng/util/parser/ExpressionNodeArithmeticOperator.java#L45
The method `calculate()` at lines 141 - 180 uses the methods add(), sub(), ...

Reply:
> Can you guarantee that exactly one of `E2<32`, `E2==32`, and `E2>32` is true?

Thanks for asking. While taking a second look at the code, I found a couple of bugs. I also added tests for the compare operatiors.

ExpressionNodeComparingOperator
https://github.com/danielb987/JMRI/blob/LogixNG_new5_006/java/src/jmri/jmrit/logixng/util/parser/ExpressionNodeComparingOperator.java

ExpressionNodeComparingOperatorTest
https://github.com/danielb987/JMRI/blob/LogixNG_new5_006/java/test/jmri/jmrit/logixng/util/parser/ExpressionNodeComparingOperatorTest.java